rfc: auxiliary leaf data in SQL abstract trees

knz opened the pull request:
Discussed last week with @jordanlewis. 

Needed for #10055.

knz committed:
rfc: auxiliary leaf data in SQL abstract trees

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18204)
<!-- Reviewable:end -->

RaduBerinde commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[docs/RFCS/sql_aux_data.md, line 226 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtHuE0Wez_cmqUJafQu:-KtHuE0Wez_cmqUJafQv:bautx9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L226)):*
> ```Markdown
>   here that this type dance is entirely *unnecessary*: from one row to
>   the next, the type of the SQL value for a given result column is
>   *always the same*. Storing then checking the Go type information in
> ```

One blemish here is `DNull` which is a separate type. So on some rows we may have `Dint` (for example), and on others `DNull`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

knz commented:
If there's an int, -1 can signal DNull.

knz commented:
And for result rows we can have a bitmap.

tschottdorf commented on docs/RFCS/sql_aux_data.md:
```diff
@@ -0,0 +1,449 @@
+
+CockroachDB currently defines 17 elementary Go types to hold SQL
+values (`DInt`, `DString`, `DInterval`, etc.), increasing to 18 when
+#18171 merges, and possibly increasing further as we extend pg
```
You might have to escape this `#`.

tbg reviewed (commented):
Informative read, thanks for the write up!

jordanlewis commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/sql_aux_data.md, line 389 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtI2lQtC77brDH8x8AQ:-KtI2lQu5DSttsMoSNw6:b-qazj6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L389)):*
> ```Markdown
> 
> - Slight deviation from 100-level data structure textbooks. (Although
>   the new propsoed pattern is not really unknown to 200- or 300-level
> ```

proposed*

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/sql_aux_data.md, line 422 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtIUDp7Upan0nIL23fY:-KtIUDp7Upan0nIL23fZ:b-97lhvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L422)):*
> ```Markdown
> 
> - do nothing: further SQL development and run-time performance slowly
>   grind to a halt.
> ```

Heh, that's a bit of hyperbole. Also, `s/grind/grinds/g`.

---

*[docs/RFCS/sql_aux_data.md, line 426 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtIa-_3D5Pa6A1HD25N:-KtIa-_3D5Pa6A1HD25O:btvrgqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L426)):*
> ```Markdown
>   which enables programmers to control allocations and avoid paying
>   run-time overhead for typing when it's known to be safe. This
>   alternative is serious but a longer-term endeavour. Also even if the
> ```

I think this actually minimizes the work involved. In addition to rewriting a ton of code, there would be significant technical obstacles involved in glueing it together with Go (efficiently). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

knz commented:
I extended the text to highlight how to treat NULLs properly. PTAL. If there are no more comments I'll start the FCP this week.

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/sql_aux_data.md, line 178 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtE7eAZExHT8UTZoEL_-r1-178:-KtfqxQY7MQ38i-tRpBu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L178)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You might have to escape this `#`.
</blockquote></details>

Done.

---

*[docs/RFCS/sql_aux_data.md, line 226 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtHuE0Wez_cmqUJafQu:-KtfrHZhYrfXIKlifdTE:b-6x11g7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L226)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

One blemish here is `DNull` which is a separate type. So on some rows we may have `Dint` (for example), and on others `DNull`.
</blockquote></details>

Added a paragraph to clarify here and below.

---

*[docs/RFCS/sql_aux_data.md, line 389 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtI2lQtC77brDH8x8AQ:-Ktfs1Gx3_npXBgPuFtc:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L389)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

proposed*
</blockquote></details>

Fixed.

---

*[docs/RFCS/sql_aux_data.md, line 422 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtIUDp7Upan0nIL23fY:-KtfsAtq_CwrbWlXKPAL:b-2ixzkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L422)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Heh, that's a bit of hyperbole. Also, `s/grind/grinds/g`.
</blockquote></details>

Thanks! Fixed, and added a note to that extent.

---

*[docs/RFCS/sql_aux_data.md, line 426 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KtIa-_3D5Pa6A1HD25N:-KtfsMsiWu79YqQiVdyH:b-3ts5go) ([raw file](https://github.com/cockroachdb/cockroach/blob/5597e94576c0e1cdbf479182bcbd2ab4a1cef402/docs/RFCS/sql_aux_data.md#L426)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think this actually minimizes the work involved. In addition to rewriting a ton of code, there would be significant technical obstacles involved in glueing it together with Go (efficiently). 
</blockquote></details>

Changed the phrasing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-:-KtfsY1A6aVc-3likghD:bjefhco)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on docs/RFCS/sql_aux_data.md:
```diff
@@ -0,0 +1,464 @@
+- when a placeholder is used multiple times (e.g. `select $1 where x >
+  $1`), the type information is stored multiple times in the AST.
+- the tree rewrite at each execute of a prepared query is expensive
+  and puts pressure on Go's heap allocator and GC.
```
sounds like a big deal!

vivekmenezes reviewed (commented):
The motivation section is very well written! Awesome doc, thanks

LGTM

knz commented:
final comment period.

couchand commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/sql_aux_data.md, line 207 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPwjdYf4q3Ux0PCGcg:-KuPwjdZ-f57x4zNDtO8:b-oz5lrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L207)):*
> ```Markdown
> 
> - `Datum` types are actually embedded in SQL expression trees using an
>   interface reference (`Datum` and/of `Expr`). This means that even
> ```

s/of/or/

---

*[docs/RFCS/sql_aux_data.md, line 228 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPxK4RQCWv4nfs_hxp:-KuPxK4SagPtpKNhVKlF:bdv5df2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L228)):*
> ```Markdown
>   *always the same* (*). Storing then checking the Go type information in
>   memory for every cell of every row in a plan's result column is
>   horrendously redundant.
> ```

Isn't this essentially the interpreter/compiler tradeoff?  Perhaps I'm missing something in this discussion, but I don't see how the SQL execution engine could be written to be both general (supports all legal statements at runtime) and optimally efficient (not require boxing primitive data types).  I suppose maybe it would involve transposing the standard row-wise computation of results into a column-wise version that allows building a query plan that operates on just the data types of that particular sequence?  Or maybe JITting?

---

*[docs/RFCS/sql_aux_data.md, line 231 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPyURA85m_pnJYI8Vq:-KuPyURA85m_pnJYI8Vr:b-fkgsuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L231)):*
> ```Markdown
> 
> (*) Regarding the handling of NULL values: NULL is a value that
> inhabits every type. That is, `NULL::int` and `NULL::string` are two
> ```

Well, it might be nice to have nullable types defined to be distinct type (or use a Maybe type constructor), to, for instance, help enforce that non-nullable columns never receive NULL as a value.  But that's not really relevant to this discussion.

---

*[docs/RFCS/sql_aux_data.md, line 246 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzCRVrDGcX_qPvz9U:-KuPzCRVrDGcX_qPvz9V:byjb1kw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L246)):*
> ```Markdown
> - the application domain strongly favors/recommends immutable trees,
>   but the logical values in the trees require replacement between
>   uses. This requires an expensive "rewrite via new allocation" traversal.
> ```

If the "logical values" change, in what sense is the tree immutable?  If I can't replace the object at the leaf node, but I can still change it's values, aren't I committing just as grave of a sin?

---

*[docs/RFCS/sql_aux_data.md, line 256 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzUbWlPUPFk3AW9Fz:-KuPzUbWlPUPFk3AW9G-:b-s5mx42) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L256)):*
> ```Markdown
>   the same column in many rows, multiple instances of the same
>   placeholders in different places of an expression) it is wasteful to
>   duplicate this type information across all instances.
> ```

I think I see what you're getting at now.  In the case of `Datum`, even primitives would still be hidden behind the interface, but the type switch would no longer be necessary?  That doesn't seem like all that big of a win, but maybe I'm missing something.

---

*[docs/RFCS/sql_aux_data.md, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzpOkcLfRbUT_jcXy:-KuPzpOkcLfRbUT_jcXz:bdwboli) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L308)):*
> ```Markdown
> is only performed at the point of use, not stored in the tree
> directly. This enables minimal data storage in the tree and efficient
> substitutions of the corresponding values without having to mutate the
> ```

As I asked above, what's the real difference between "substitution" and "mutation" here?  Won't the same safety/bug questions arise from either strategy?

---

*[docs/RFCS/sql_aux_data.md, line 342 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuQ-Q5cnGspnT967bkd:-KuQ-Q5cnGspnT967bke:bty8vfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L342)):*
> ```Markdown
>    it is time to drop the idea to name placeholders and instead number
>    them, which will in turn make the data structure even smaller and
>    more efficient to use (lookups using an array instead of a map).
> ```

Do the placeholders have to be dense?  In practice they probably are.  But it seems worth noting the difference between the current enforcement of numeric identifiers and enforcing the smallest possible numeric identifiers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/sql_aux_data.md, line 207 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPwjdYf4q3Ux0PCGcg:-KuwLYrerBKn2tVINL7s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L207)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

s/of/or/
</blockquote></details>

Done.

---

*[docs/RFCS/sql_aux_data.md, line 228 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPxK4RQCWv4nfs_hxp:-KuwLizrSZDJ9IFkjqmH:b-q06uhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L228)):*
> Isn't this essentially the interpreter/compiler tradeoff? 

No, it's about choosing the right data structures. This always pays off regardless of the evaluation strategy of the language itself.

> Perhaps I'm missing something in this discussion, but I don't see how the SQL execution engine could be written to be both general (supports all legal statements at runtime) and optimally efficient (not require boxing primitive data types).  

You're right however I don't think this is relevant in this discussion.

> I suppose maybe it would involve transposing the standard row-wise computation of results into a column-wise version [...]

Yes eventually this is where the correct solution to that particular instance of the general problem is to be found. We have a roadmap item to go there eventually.

---

*[docs/RFCS/sql_aux_data.md, line 231 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPyURA85m_pnJYI8Vq:-KuwMEOTqx5Ic6JIXJzy:b-pbdqzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L231)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Well, it might be nice to have nullable types defined to be distinct type (or use a Maybe type constructor), to, for instance, help enforce that non-nullable columns never receive NULL as a value.  But that's not really relevant to this discussion.
</blockquote></details>

Agreed!

---

*[docs/RFCS/sql_aux_data.md, line 246 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzCRVrDGcX_qPvz9U:-KuwMJ_hq8Kn2j26eHtf:bonancw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L246)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

If the "logical values" change, in what sense is the tree immutable?  If I can't replace the object at the leaf node, but I can still change it's values, aren't I committing just as grave of a sin?
</blockquote></details>

The issue is that the data structure instances in memory are/will be shared across multiple versions. E.g. with AST caching we want to reuse the tree data as much as possible even though each particular execution needs to rewrite the tree.

---

*[docs/RFCS/sql_aux_data.md, line 256 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzUbWlPUPFk3AW9Fz:-KuwMaQjvJRjfnw6IibA:brfo4qt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L256)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think I see what you're getting at now.  In the case of `Datum`, even primitives would still be hidden behind the interface, but the type switch would no longer be necessary?  That doesn't seem like all that big of a win, but maybe I'm missing something.
</blockquote></details>

Eventually the Datum interface would disappear.

---

*[docs/RFCS/sql_aux_data.md, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzpOkcLfRbUT_jcXy:-KuwMfLUhoU66ovmOQ6x:b-mno1mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L308)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

As I asked above, what's the real difference between "substitution" and "mutation" here?  Won't the same safety/bug questions arise from either strategy?
</blockquote></details>

Tree updates by substitution allow sharing the nodes between multiple versions of the tree, see above.

---

*[docs/RFCS/sql_aux_data.md, line 342 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuQ-Q5cnGspnT967bkd:-KuwMmv_nNHyEQKas07L:bntyvi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/sql_aux_data.md#L342)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Do the placeholders have to be dense?  In practice they probably are.  But it seems worth noting the difference between the current enforcement of numeric identifiers and enforcing the smallest possible numeric identifiers.
</blockquote></details>

Added a sentence to clarify.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

couchand commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 246 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzCRVrDGcX_qPvz9U:-Kuyf-FzUxTbvDSvftgB:b-3enqi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/20170926_sql_aux_data.md#L246)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The issue is that the data structure instances in memory are/will be shared across multiple versions. E.g. with AST caching we want to reuse the tree data as much as possible even though each particular execution needs to rewrite the tree.
</blockquote></details>

I think I still take issue with considering this to still be an immutable tree.  We can actively make the performance/purity tradeoff, because sometimes it makes sense to sacrifice purity for the sake of performance.  But let's not pretend like we're not sacrificing purity.  And as long as we're at it, let's just go ahead and mutate the tree for other types of changes we're making.  I fail to see how splitting the difference here (unchanging tree structure with changing leaf values) maintains any of the bug-squashing advantages of immutable trees anymore.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 256 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzUbWlPUPFk3AW9Fz:-Kuyfgly2FHN81t4wsy0:b-5bs1jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/20170926_sql_aux_data.md#L256)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Eventually the Datum interface would disappear.
</blockquote></details>

I think my confusion here is linked to my comment above about the compiler/interpreter divide.  If there's no general interface, doesn't that demand compilation?

---

*[docs/RFCS/20170926_sql_aux_data.md, line 308 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzpOkcLfRbUT_jcXy:-KuyfxKC3qflCc5f5vTI:b-apayz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/20170926_sql_aux_data.md#L308)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Tree updates by substitution allow sharing the nodes between multiple versions of the tree, see above.
</blockquote></details>

See my comment above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 246 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzCRVrDGcX_qPvz9U:-Kuz3LpUsnUstx09_QtZ:b-y74w1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/20170926_sql_aux_data.md#L246)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think I still take issue with considering this to still be an immutable tree.  We can actively make the performance/purity tradeoff, because sometimes it makes sense to sacrifice purity for the sake of performance.  But let's not pretend like we're not sacrificing purity.  And as long as we're at it, let's just go ahead and mutate the tree for other types of changes we're making.  I fail to see how splitting the difference here (unchanging tree structure with changing leaf values) maintains any of the bug-squashing advantages of immutable trees anymore.
</blockquote></details>

Please use more words in your discussion here. What is the problem you are trying to solve?
1) this RFC acknowledge that the current code is heavy in tree mutations (either by substitution, or erroneously in-place sometimes), and overall proposes to do away with that so the trees become properly immutable. Are you suggesting this proposal is not useful?
2) the trade-off between in-place mutation and update-by-copy is extremely well-known and practiced either one way or another in every language/library under the sun. In CockroachDB we made the choice to avoid in-place mutations for a variety of reasons that are still valid but outside of the scope of the discussion. What do you want me to add here?

---

*[docs/RFCS/20170926_sql_aux_data.md, line 256 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzUbWlPUPFk3AW9Fz:-Kuz3znYjo7GsZ4HIuSV:bt3yoxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/20170926_sql_aux_data.md#L256)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think my confusion here is linked to my comment above about the compiler/interpreter divide.  If there's no general interface, doesn't that demand compilation?
</blockquote></details>

Not necessarily. You can also  eliminate dynamic dispatch by specializing (parts of) the traversal code to the value types. Also, note that this RFC does not aim to propose the technical mechanisms to achieve those things, and merely states we need to change the data structure upfront to enable further work on the solution(s).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

couchand commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 246 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzCRVrDGcX_qPvz9U:-KuzE9NHLFbrL2wFuvcr:bdoof47) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/20170926_sql_aux_data.md#L246)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please use more words in your discussion here. What is the problem you are trying to solve?
1) this RFC acknowledge that the current code is heavy in tree mutations (either by substitution, or erroneously in-place sometimes), and overall proposes to do away with that so the trees become properly immutable. Are you suggesting this proposal is not useful?
2) the trade-off between in-place mutation and update-by-copy is extremely well-known and practiced either one way or another in every language/library under the sun. In CockroachDB we made the choice to avoid in-place mutations for a variety of reasons that are still valid but outside of the scope of the discussion. What do you want me to add here?
</blockquote></details>

I mean to say, either strategy is fine, but we should pick one and commit to it.  I agree that there are lots of benefits to avoiding mutations.  It seems like, to actually realize those benefits, you have to be strict about preventing mutations anywhere (a point that seems to be echoed by your comment "erroneously in-place sometimes").  I don't think I understand how requiring the structure to be immutable but allowing the leaf values to be mutable is sufficiently strict that it will afford the advantages of treating the trees as immutable.  Specifically, wouldn't a change ("substitution") to a leaf value "pull the rug out" from under the assumption that the tree won't change?  How is it that a substitution doesn't count as a mutation?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 246 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KuPzCRVrDGcX_qPvz9U:-KuzRjyRFlLVLmlCouXJ:b-jvqbms) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dfe2550ef2f35cab18c9bb2e8dcf0cb87a46476/docs/RFCS/20170926_sql_aux_data.md#L246)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I mean to say, either strategy is fine, but we should pick one and commit to it.  I agree that there are lots of benefits to avoiding mutations.  It seems like, to actually realize those benefits, you have to be strict about preventing mutations anywhere (a point that seems to be echoed by your comment "erroneously in-place sometimes").  I don't think I understand how requiring the structure to be immutable but allowing the leaf values to be mutable is sufficiently strict that it will afford the advantages of treating the trees as immutable.  Specifically, wouldn't a change ("substitution") to a leaf value "pull the rug out" from under the assumption that the tree won't change?  How is it that a substitution doesn't count as a mutation?
</blockquote></details>

Oh now I see what you mean.
So the goal of this entire RFC is to propose that the actual *values* of leaf data do not live in the tree directly. In particular, the decision of *what is the particular value* of a leaf item (i.e. what is the result of its "evaluation") should not be a property of the tree any more, but a property of the evaluation context. Once this is true, we can change the value of evaluating a tree (by changing the evaluation of leaf data) without having to modify the tree any more. Does this clarify? Perhaps I should add a sentence to that effect, let me know what works for you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM. This makes a lot of sense to me. In addition to the performance concerns, the "kind of immutable but not really" tree we have now is error-prone (e.g. we rely that all code plays nice and makes modifications that don't affect a retry of the statement).

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 233 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KvXJg6E0II4iN5Yhe6I:-KvXJg6E0II4iN5Yhe6J:bk7txgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7acc21dbe149ef29c8478346f9ddafbd929a3fe/docs/RFCS/20170926_sql_aux_data.md#L233)):*
> ```Markdown
> inhabits every type. That is, `NULL::int` and `NULL::string` are two
> different things in SQL. Even in a column where NULL *values* can
> occur, all values in the columns are still of the same type.
> ```

I agree with this at a high level, but it's not the case in the current implementation (all values in a column are not of the same *Go* type).

I see the proposal below does describe handling of NULLs, so maybe here we should just say here that NULLs complicate things a bit but the proposal below addresses the issue.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 317 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KvXKtDVmTNQl5WXTeLO:-KvXKtDVmTNQl5WXTeLP:b-piu038) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7acc21dbe149ef29c8478346f9ddafbd929a3fe/docs/RFCS/20170926_sql_aux_data.md#L317)):*
> ```Markdown
> | Current code | New code | Notes |
> |--------------|----------|-------|
> | `type IndexedVar struct {...}` | `type IndexedVar int` | |
> ```

[nit] We may want to rename these as `IndexedVarIdx`, `PlaceholderIdx` etc

---

*[docs/RFCS/20170926_sql_aux_data.md, line 382 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KvXLyr-ZDOiuObYw50S:-KvXLyr-ZDOiuObYw50T:b-1fcakr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7acc21dbe149ef29c8478346f9ddafbd929a3fe/docs/RFCS/20170926_sql_aux_data.md#L382)):*
> ```Markdown
> measures/decides which one to keep.
> 
> This strongly suggest to implement the link between logical plans not
> ```

[nit] suggests

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-:-KvXMEwPR_pdS0BtjDSJ:b-1gr3fr)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 233 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KvXJg6E0II4iN5Yhe6I:-Kw53GH7c-josFx8KtuO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7acc21dbe149ef29c8478346f9ddafbd929a3fe/docs/RFCS/20170926_sql_aux_data.md#L233)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I agree with this at a high level, but it's not the case in the current implementation (all values in a column are not of the same *Go* type).

I see the proposal below does describe handling of NULLs, so maybe here we should just say here that NULLs complicate things a bit but the proposal below addresses the issue.
</blockquote></details>

Done.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 317 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KvXKtDVmTNQl5WXTeLO:-Kw53Suvl0ur84K9Ec9w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7acc21dbe149ef29c8478346f9ddafbd929a3fe/docs/RFCS/20170926_sql_aux_data.md#L317)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] We may want to rename these as `IndexedVarIdx`, `PlaceholderIdx` etc
</blockquote></details>

Done.

---

*[docs/RFCS/20170926_sql_aux_data.md, line 382 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204#-KvXLyr-ZDOiuObYw50S:-Kw53VgxWnJxi-aeEm3f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7acc21dbe149ef29c8478346f9ddafbd929a3fe/docs/RFCS/20170926_sql_aux_data.md#L382)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] suggests
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18204)*
<!-- Sent from Reviewable.io -->
